src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs

```diff
@@ -0,0 +1,146 @@
+        string snapshotUrl = snapshotConfig.DownloadUrl ??
+                             throw new InvalidOperationException("Snapshot download URL is not configured");
+        byte[]? snapshotChecksum =
+            snapshotConfig.Checksum is null ? null : Bytes.FromHexString(snapshotConfig.Checksum);
+
+        string snapshotFileName = Path.GetTempFileName();
+
+        await DownloadSnapshotTo(snapshotUrl, snapshotFileName, cancellationToken);
+
+        if (snapshotChecksum is not null)
+        {
+            bool isChecksumValid = await VerifyChecksum(snapshotFileName, snapshotChecksum);
+            if (!isChecksumValid)
+            {
+                if (_logger.IsError)
+                    _logger.Error("Snapshot checksum verification failed. Aborting, but will continue running.");
+                return;
+            }
+
+            if (_logger.IsInfo)
+                _logger.Info("Snapshot checksum verified.");
+        }
+        else if (_logger.IsWarn)
+            _logger.Warn("Snapshot checksum is not configured");
+
+
+        await ExtractSnapshotTo(snapshotFileName, dbPath, cancellationToken);
+
+        if (_logger.IsInfo)
+            _logger.Info("Database successfully initialized from snapshot.");
+    }
+
+    private async Task DownloadSnapshotTo(
+        string snapshotUrl, string snapshotFileName, CancellationToken cancellationToken)
+    {
+        if (_logger.IsInfo)
+            _logger.Info($"Downloading snapshot from {snapshotUrl}");
+
+        using HttpClient httpClient = new();
+
+        HttpRequestMessage request = new(HttpMethod.Get, snapshotUrl);
+        FileInfo snapshotFileInfo = new(snapshotFileName);
+        if (snapshotFileInfo.Exists)
+            request.Headers.Range = new RangeHeaderValue(snapshotFileInfo.Length, null);
+
+        using HttpResponseMessage response =
+            (await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken))
+            .EnsureSuccessStatusCode();
+
+        await using Stream contentStream = await response.Content.ReadAsStreamAsync(cancellationToken);
+        await using FileStream snapshotFileStream = new(
+            snapshotFileName, FileMode.Append, FileAccess.Write, FileShare.None, BufferSize, true);
+
+        long totalBytesRead = snapshotFileInfo.Length;
+        long? totalBytesToRead = response.Content.Headers.ContentLength;
+
+        using ProgressTracker progressTracker = new(
+            _api.LogManager, _api.TimerFactory, TimeSpan.FromSeconds(5), totalBytesRead, totalBytesToRead);
+
+        byte[] buffer = new byte[BufferSize];
+        while (true)
+        {
+            int bytesRead = await contentStream.ReadAsync(buffer, cancellationToken);
+            if (bytesRead == 0)
+                break;
+            await snapshotFileStream.WriteAsync(buffer.AsMemory(0, bytesRead), cancellationToken);
+
+            progressTracker.AddProgress(bytesRead);
+        }
```

Comment:
According to https://stackoverflow.com/a/230141/11035317 this is not the fastest way of doing it. Is this something we should focus on?

Reply:
the problem with the CopyTo method is that it doesn't allow me to track progress. Maybe we could wrap the contentStream so we can count read bytes 🤔

Reply:
I think [what they do in ``CopyToAsync``](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs,87) is similar to what is done here already

Reply:
yes, I was looking into it and I think the same